src/plan_agg_bookend.c

```diff
@@ -0,0 +1,718 @@
+#include "catalog/pg_proc.h"
+#include <catalog/namespace.h>
+#include "utils/typcache.h"
+#include "access/stratnum.h"
+#include "plan_agg_bookend.h"
+#include "utils.h"
+
+typedef struct FirstLastAggInfo
+{
+	MinMaxAggInfo *m_agg_info;	/* reusing MinMaxAggInfo to avoid code
+								 * duplication */
+	Expr	   *sort;			/* Expression to use for ORDER BY */
+} FirstLastAggInfo;
+
+typedef struct MutatorContext
+{
+	MinMaxAggPath *mm_path;
+} MutatorContext;
+
+/*  static const char FIRST[] = "first"; */
+/*  static const char LAST[] = "last"; */
+
+static bool find_first_last_aggs_walker(Node *node, List **context);
+static bool build_first_last_path(PlannerInfo *root, FirstLastAggInfo *flinfo,
+					  Oid eqop, Oid sortop, bool nulls_first);
+static void first_last_qp_callback(PlannerInfo *root, void *extra);
+static Node *mutate_aggref_node(Node *node, MutatorContext *context);
+static void replace_aggref_in_tlist(MinMaxAggPath *minmaxagg_path);
+
+/*
+ * mutate_aggref_node
+ *
+ * Mutator function used by recursive `expression_tree_mutator`
+ * to replace Aggref node with Param node
+ */
+Node *
+mutate_aggref_node(Node *node, MutatorContext *context)
+{
+	if (node == NULL)
+		return NULL;
+	if (IsA(node, Aggref))
+	{
+		Aggref	   *aggref = (Aggref *) node;
+
+		/* See if the Aggref should be replaced by a Param */
+		if (context->mm_path != NULL &&
+			list_length(aggref->args) == 2)
+		{
+			TargetEntry *curTarget = (TargetEntry *) linitial(aggref->args);
+			ListCell   *cell;
+
+			foreach(cell, context->mm_path->mmaggregates)
+			{
+				MinMaxAggInfo *mminfo = (MinMaxAggInfo *) lfirst(cell);
+
+				if (mminfo->aggfnoid == aggref->aggfnoid &&
+					equal(mminfo->target, curTarget->expr))
+					return (Node *) copyObject(mminfo->param);
+			}
+		}
+	}
+	return expression_tree_mutator(node, mutate_aggref_node, (void *) context);
+}
+
+/*
+ * replace_aggref_in_tlist
+ *
+ * If MinMaxAggPath is chosen, instead of running aggregate
+ * function we will execute subquery that we've generated. Since we
+ * use subquery we need to replace target list Aggref node with Param
+ * node. Param node passes output value from the subquery.
+ *
+ */
+void
+replace_aggref_in_tlist(MinMaxAggPath *minmaxagg_path)
+{
+	MutatorContext context;
+
+	context.mm_path = minmaxagg_path;
+
+	((Path *) minmaxagg_path)->pathtarget->exprs = (List *) mutate_aggref_node((Node *) ((Path *) minmaxagg_path)->pathtarget->exprs, (void *) &context);
+}
+
+/* Stores function id (FIRST/LAST) with proper comparison strategy */
+typedef struct FuncStrategy
+{
+	Oid			func_oid;
+	StrategyNumber strategy;
+} FuncStrategy;
+
+static Oid	first_last_arg_types[] = {ANYELEMENTOID, ANYOID};
+
+struct FuncStrategy first_func_strategy = {.func_oid = InvalidOid,.strategy = BTLessStrategyNumber};
+struct FuncStrategy last_func_strategy = {.func_oid = InvalidOid,.strategy = BTGreaterStrategyNumber};
+
+static Oid
+get_function_oid(char *name, int nargs, Oid arg_types[])
```

Comment:
is this a repetition of another similar func in another place. Should we move this into utils?